protocols/upnp/src/provider.rs

```diff
@@ -0,0 +1,189 @@
+                                    || matches!(ip.segments(), [0x100, 0, 0, 0, _, _, _, _])
+                                    // IETF Protocol Assignments (`2001::/23`)
+                                    || (matches!(ip.segments(), [0x2001, b, _, _, _, _, _, _] if b < 0x200)
+                                    && !(
+                                    // Port Control Protocol Anycast (`2001:1::1`)
+                                    u128::from_be_bytes(ip.octets()) == 0x2001_0001_0000_0000_0000_0000_0000_0001
+                                    // Traversal Using Relays around NAT Anycast (`2001:1::2`)
+                                    || u128::from_be_bytes(ip.octets()) == 0x2001_0001_0000_0000_0000_0000_0000_0002
+                                    // AMT (`2001:3::/32`)
+                                    || matches!(ip.segments(), [0x2001, 3, _, _, _, _, _, _])
+                                    // AS112-v6 (`2001:4:112::/48`)
+                                    || matches!(ip.segments(), [0x2001, 4, 0x112, _, _, _, _, _])
+                                    // ORCHIDv2 (`2001:20::/28`)
+                                    || matches!(ip.segments(), [0x2001, b, _, _, _, _, _, _] if (0x20..=0x2F).contains(&b))
+                                ))
+                                    // code for Ipv4::is_documentation()
+                                    || (ip.segments()[0] == 0x2001) && (ip.segments()[1] == 0xdb8)
+                                    // code for Ipv4::is_unique_local()
+                                    || (ip.segments()[0] & 0xfe00) == 0xfc00
+                                    // code for Ipv4::is_unicast_link_local()
+                                    || (ip.segments()[0] & 0xffc0) == 0xfe80)
+        }
+    }
+}
+
+/// Interface that interacts with the inner gateway by messages,
+/// `GatewayRequest`s and `GatewayEvent`s.
+pub struct Gateway {
+    pub(crate) sender: Sender<GatewayRequest>,
+    pub(crate) receiver: Receiver<GatewayEvent>,
+    pub(crate) external_addr: IpAddr,
+}
+
+/// Abstraction to allow for compatibility with various async runtimes.
+#[async_trait]
+pub trait Provider {
+    async fn search_gateway() -> Result<Gateway, Box<dyn Error>>;
+}
+
+macro_rules! impl_provider {
+    ($impl:ident, $executor: ident, $gateway:ident, $protocol: ident) => {
+        use super::Gateway;
+        use crate::behaviour::{GatewayEvent, GatewayRequest};
+
+        use async_trait::async_trait;
+        use futures::{channel::mpsc, SinkExt, StreamExt};
+        use igd_next::SearchOptions;
+        use std::error::Error;
+
+        #[async_trait]
+        impl super::Provider for $impl {
+            async fn search_gateway() -> Result<super::Gateway, Box<dyn Error>> {
+                let gateway = $gateway::search_gateway(SearchOptions::default()).await?;
+                let external_addr = gateway.get_external_ip().await?;
```

Comment:
This is still IO work, correct? I think this shouldn't be on the main task because it will be polled by the behaviour.

Reply:
The easiest will probably be to spawn a task here and simply await that task. It feels unnecessary but it is much cheaper than doing the actual IO here.

Reply:
yup good catch Thomas! Updated to use a `oneshot` instead, seemed simpler and more elegant as with spawning a task we have the required abstraction with the executor. Ptal!